### PR TITLE
feat: draw-pattern section

### PR DESCRIPTION
**Pull request by BQXBQX**
# feat: use section component && use zustand
image

### CODE REVIEW

**Comment by BQXBQX** [comment]
## Complete node lazy loading optimization
Here is the difference in rendering time when dragging a node when the page has 100 nodes:
image
Screenshot 2024-08-17 at 18 05 17

**Comment by BQXBQX** [comment]
## Complete the type classification and group the types in the types folder
image

**Comment by BQXBQX** [comment]
# Issue
After calling the graph-editor package packaged with `vite`, the canvas cannot be dragged and the newly added nodes are not visible. However, when exporting the image, the added nodes are on the image. There is no problem using the component `index.tsx` directly without `vite` packaging

## `vite config`
```typescript
import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';
import path from 'path';

export default defineConfig(({ mode }) => {
  if (mode === 'production') {
    return {
      build: {
        lib: {
          entry: path.resolve(__dirname, 'src/index.tsx'),
          name: '@graphscope/studio-graph-editor',
          fileName: format => `studio-graph-editor.${format}.js`,
        },
        outDir: './dist',
        rollupOptions: {
          external: ['react', 'react-dom'],
        },
      },
      resolve: {
        dedupe: ['react', 'react-dom'],
      },
      plugins: [react()],
    };
  } else {
    return {
      root: './',
      server: {
        port: 8000,
        open: '/',
      },
      build: {
        outDir: './dist',
      },
      plugins: [react()],
    };
  }
});
```

## `package.json`
```json
{
  "name": "@graphscope/studio-graph-editor",
  "version": "0.1.0",
  "description": "",
  "main": "./dist/studio-graph-editor.umd.js",
  "module": "./dist/studio-graph-editor.es.js",
  "types": "./dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/GraphScope/portal.git"
  },
  "files": [
    "es",
    "lib",
    "dist"
  ],
  "scripts": {
    "start": "vite dev",
    "build": "vite build && tsc"
  },
  "peerDependencies": {
    "react": "18.2.0",
    "react-dom": "18.2.0"
  },
  "dependencies": {
    "@graphscope/studio-components": "workspace:*",
    "antd": "^5.17.0",
    "d3-force": "latest",
    "dagre": "latest",
    "html-to-image": "^1.11.11",
    "lodash": "^4.17.21",
    "react-intl": "^6.6.1",
    "reactflow": "latest",
    "rxjs": "^7.8.1",
    "uuid": "^9.0.1",
    "valtio": "2.0.0-rc.1",
    "zustand": "^4.5.5"
  },
  "devDependencies": {
    "@types/d3-force": "latest",
    "@types/lodash": "^4.14.202",
    "@vitejs/plugin-react": "^4.2.1",
    "vite": "^5.0.12"
  },
  "author": "",
  "license": "ISC"
}
```
after changing module and main to `./src/index.tsx`, no problem and normal operation.

**Comment by BQXBQX** [comment]
## Complete the basic UI and layout of the page
image

**Comment by BQXBQX** [comment]
# Complete popover content
image

**Comment by BQXBQX** [comment]
## Complete cypher sentence generate 
image

**Comment by BQXBQX** [comment]
## Complete Modal
image

**Comment by BQXBQX** [comment]
## Add page script
image

File: packages/studio-draw-pattern/src/components/Canvas/PopoverContent.tsx [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    {
+      title: 'Compare',
+      dataIndex: 'compare',
+      render: (_: any, record: Property) => (
+        <Select
+          defaultValue={record.compare}
+          onChange={value => handleChange(record.id, 'compare', value)}
+          style={{ width: '100%' }}
+          options={[
+            { value: '>', label: '>' },
```
Comment (pomelo-nwu) [inline packages/studio-draw-pattern/src/components/Canvas/PopoverContent.tsx:62]: 操作符应该还有多个

File: packages/studio-draw-pattern/src/components/DrawPattern.tsx [hunk 2]
```diff
@@ -1,4 +1,59 @@
-import React from 'react';
-export const DrawPattern = () => {
-  return <div>hello</div>;
+import React, { createContext } from 'react';
+import { QuickStart } from './QuickStart';
+import { Canvas } from './Canvas';
+import { Section } from '@graphscope/studio-components';
+import { ISchemaNode } from '@graphscope/studio-graph-editor/dist/types/node';
```
Comment (pomelo-nwu) [inline packages/studio-draw-pattern/src/components/DrawPattern.tsx:5]: 这个地方的 import url 不对
应该import type { ISchemaNode } from '@graphscope/studio-graph-editor'

File: packages/studio-draw-pattern/src/components/QuickStart/QuickStartItem.tsx [hunk 3]
```diff
@@ -1,3 +1,68 @@
...
+  const handleClick = useCallback(() => {
+    // 每次 selection change 都要清空 store
+    clearNode && clearNode();
+    clearEdge && clearEdge();
+    clearGraphStore();
+    clearProperties();
+
+    // 根据 template id 获取模板数据
+    switch (id) {
+      case 'triangle-loop':
```
Comment (pomelo-nwu) [inline packages/studio-draw-pattern/src/components/QuickStart/QuickStartItem.tsx:32]: 模版应该更多一点，同时考虑 switch case 的扩展性

File: packages/studio-draw-pattern/src/data.ts [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+      label: 'Challenge',
+      primary: 'id',
+      saved: true,
+      disabled: true,
+      properties: [
+        {
+          key: '2537c7e6-0bd5-447c-b651-0fac1482f4a7',
+          index: 0,
+          token: '',
+          name: 'id',
```
Comment (pomelo-nwu) [inline packages/studio-draw-pattern/src/data.ts:14]: 这里的属性key，似乎没有太多意义？

File: packages/studio-draw-pattern/src/utils/encode.ts [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  });
+};
+
+export const encodeNodes = (nodes: Node[], encodeSingleNodeCallback: (preNode: Node, currentNode: Node) => void) => {
+  if (!nodes) return;
+
+  nodes
+    .filter(node => node !== undefined)
+    .forEach(node => {
+      const newStatement = node.data?.data?.label && `:${node.data.data.label}`;
```
Comment (pomelo-nwu) [inline packages/studio-draw-pattern/src/utils/encode.ts:36]: 这里的数据结构似乎有点嵌套太深？（这条不是修改建议）

File: packages/studio-draw-pattern/test/utils/encodeEdges.test.ts [hunk 6]
```diff
@@ -0,0 +1,240 @@
+import { describe, it, expect, vi } from 'vitest';
+import { Edge } from '../../src/types/edge';
+import { encodeEdges } from '../../src/utils/encode';
+
+describe('encodeEdges', () => {
+  it('should encode a single edge correctly', () => {
+    const edges: Edge[] = [
+      {
+        edgeKey: 'e1',
+        isErgodic: false,
```
Comment (pomelo-nwu) [inline packages/studio-draw-pattern/test/utils/encodeEdges.test.ts:10]: 这个字段的含义？

**Comment by pomelo-nwu** [comment]
@BQXBQX Thanks for your contribution! It's great work. We really appreciate it, looking forward to more from you!

**Pull request merged**